### DISCUSSION TITLE
Suggestions for reducing unnecessary Scavenge Garbage Collector calls, since GC is blocking code quite many times.

### DISCUSSION THREAD

**Discussion by mai1x9** [nodejs/help#5032] [Q&A] [answered]
Hi there, 
**Could anyone please suggest on how to reduce frequent GC calls made by node (by writing better code, decreasing object allocations etc..).?** 

Looks like Scavenge GC calls are pretty fast and occurs more often than mark sweep, however, more number of GC calls made by Scavenge is blocking the code. After some research, I could understand that scavenge GC is called when max semi space allocated exceeds. I believe lesser memory allocations would less often call GC. 

-------

**Regarding Application:**
It is backend service using express. On LoadTesting with Concurrent users of 1000, with 20K requests,I could see GC calls by Scavenge is occurring very often and blocking the event loop, hence request latency is increased.

**Would like to know any suggestions**, like How to reduce unnecessary memory allocations, optimising loops where objects created in loop might be responsible for high memory raise, etc etc..

-----

**On Side Note,** 
- I the application is tested for memory leaks, as such there wasn't any memory leak, as heap used is resetting to stable 100 Mb  after Mark sweep GC call(initial value of heap used when application ran). 
- I have changed max semi space to 64 MB, but that have result in much worst results, since scavenge operation is now taking more time than default, when gc call happens.

Is it a known fact that under excessive loadtesting of server, memory does increase and very often GC calls happen?

**Comment by bnoordhuis** [comment, marked as answer]
You don't know at this point how much CPU time is spent inside the garbage collector so my suggestion is to profile your application first.

**Comment by mai1x9** [comment]
Hi @bnoordhuis after cpu profiling (from browser dev tools and running node with `--inspect`), I could get cpu timespent by GC. Regarding the loadtest, 200 concurrent users and 5k requests sent to server build using express framework. I am enclosing the results (After sorting out cpu time),

<img width="1565" alt="Screenshot 2023-01-21 at 1 21 45 PM" src="https://user-images.githubusercontent.com/43678329/213855485-7b5d8775-049c-4605-91d0-76d659ff0b00.png">

Looks like GC took around 1 second cpu time. I wonder if these **results are per requests** or cumulative time for all 5k requests during the course of run. **Can I say that GC is not effecting the application performance badly and is negligible.?**

**Reply by bnoordhuis** [reply]
That's cumulative and 2.4% is indeed quite insignificant.

**Reply by mai1x9** [reply]
@bnoordhuis thanks for the clarification. I would mark it as answered. Out of curiosity,  why does processTicksAndRejections and callbackTrampoline takes lot of cpu utilisation. 

**Reference:**
<img width="958" alt="Screenshot 2023-01-21 at 4 00 10 PM" src="https://user-images.githubusercontent.com/43678329/213873948-0bdd9017-1f48-4c50-9bd0-74de9a24c167.png">

**Is it the node internals using cpu% or issue in the code, where setTimeouts, or Promises/callbacks in the code is doing cpu intensive tasks**. I tried to search some blogs but no clear cut explanation. Would be great if you can clarify on this?

**Reply by bnoordhuis** [reply]
You've sorted by total time, not self time. Total time = self time + self time of callees. It's not processTicksAndRejections() and callbackTrampoline() that take up a lot of CPU time but the callbacks/promises they invoke.

**Comment by mai1x9** [comment]
@bnoordhuis please refer to below images, looks like gc is using high cpu % (around 65%+ too recorded), and high gc is recorded when server has no load, is it because server has no load, max time appears to be spend on gc than processing requests and hence higher cpu %? Is it because of mark sweep run ? The results captured are after load testing and i could observe mark sweep was run (using gc-stats npm module). **Just what to confirm ok, since it would be just a spike during profiling was done?**

<img width="948" alt="Screenshot 2023-01-22 at 1 00 07 PM" src="https://user-images.githubusercontent.com/43678329/213905446-55225f36-cd7b-4de3-87f4-5d279088345d.png">
<img width="1065" alt="Screenshot 2023-01-22 at 12 57 11 PM" src="https://user-images.githubusercontent.com/43678329/213905451-1227bd6c-d447-4a1f-a23b-1ea40acd116f.png">

**Reply by bnoordhuis** [reply]
In your screenshot GC looks outsized in isolation but it's only 1% of wall clock time. Nothing to lose sleep over. :-)

**Comment by bnoordhuis** [comment]
@mai1x9 You marked one of your questions as the answer but, for the benefit of people coming here from search engines, you probably want to pick one of my replies.

**Reply by mai1x9** [reply]
sorry for that. I have marked one of your answers.
